chore(blog): adding blog post "Integrate traces from vendor SDKs in your OpenTelemetry stack" (#4334)
open-telemetry/opentelemetry.io: content/en/blog/2024/integrate-datadog-traces-into-otel-stack/index.md

Concatenating this `Upper TraceID` with Datadog’s TraceID in hexadecimal
representation (that we will call `Lower TraceID` ), we have the exact TraceID
representation for an OpenTelemetry stack:

```mermaid
%%{init: {"flowchart": {"htmlLabels": false}} }%%
graph TD
UpperTraceID["`**_dd.p.tid (hex)** f3c18530c08e00a4`"]
LowerTraceID["`**dd trace id (hex)** 3b881a9ce0197d39`"]
TraceID["`**OTel Trace ID** f3c18530c08e00a43b881a9ce0197d39`"]

UpperTraceID --> TraceID
LowerTraceID --> TraceID
```

Now, to create this representation, we will reconstruct the TraceID at the
OpenTelemetry Collector level, using the `[transform
processor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/transformprocessor)`.

### Reconstructing TraceID for Datadog Spans

The OTel Collector `[transform
processor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/transformprocessor)`
is a component that allows for the transformation of span data as it passes
through the OpenTelemetry Collector. It can modify attributes of a span such as
name, kind, attributes, resources, and instrumentation library, among others.

> **In this context, it is used to reconstruct the TraceID for Datadog spans to
> facilitate a unified tracing environment.**

Reviewer: ```suggestion
In this context, the `transform processor` is used to reconstruct the TraceID for Datadog spans to
facilitate a unified tracing environment.
```

> We will declare it in the `processors` section in the OTel Collector Config with
> the OTel Collector transform processor.

Reviewer: ```suggestion
In the OTel Collector config, you can configure the `transform processor` as follows:
```

```yaml
processors:
  # ...

  transform:
    trace_statements:
      - context: span
        statements:
          # transformation statements
          - set(cache["upper_trace_id"], attributes["_dd.p.tid"]) where
            attributes["_dd.p.tid"] != nil
          - set(cache["lower_trace_id"], Substring(trace_id.string, 16, 16))
            where cache["upper_trace_id"] != nil
          - set(cache["combined_trace_id"], Concat([cache["upper_trace_id"],
            cache["lower_trace_id"]],"")) where cache["upper_trace_id"] != nil
          - set(trace_id.string, cache["combined_trace_id"]) where
            cache["combined_trace_id"] != nil
```

> And then we will add four transformation statements:

Reviewer: ```suggestion
Next, add four transformation statements:
```

- `set(cache["upper_trace_id"], attributes["_dd.p.tid"]) where attributes["_dd.p.tid"] != nil`,
  where we will capture the Upper TraceID from the `_dd.p.tid` attribute if it
  is defined and set it to a temporary cache;
- `set(cache["lower_trace_id"], Substring(trace_id.string, 16, 16)) where cache["upper_trace_id"] != nil`,
  where we will convert a Datadog TraceID into hexadecimal format (where we will
  have sixteen zeros plus the Lower TraceID). We will remove the "zeros” segment
  and grab the second half of it (this is why we have a sub string getting only
  the last 16 hex digits from the TraceID);
- `set(cache["combined_trace_id"], Concat([cache["upper_trace_id"], cache["lower_trace_id"]],"")) where cache["upper_trace_id"] != nil`,
  where we concatenate `upper_trace_id` and `lower_trace_id` into the
  OpenTelemetry TraceID;
- `set(trace_id.string, cache["combined_trace_id"]) where cache["combined_trace_id"] != nil`
  and finally we set the `trace_id` for this span with the string concatenation.

> To see it working, let’s run our example again, with a different

Reviewer: ```suggestion
To see it working, run your example again, with a different
```

`docker compose` file and then execute the same API call as before:

```bash
> docker compose -f ./docker-compose.step2.yaml up -d

> curl http://localhost:8080/remotehello
# it returns:
# {"hello":"world!","from":"quick-start-api-otel","to":"quick-start-api-datadog"}
```

> When we enter into the Jaeger UI we see that our problem was partially solved.
> Now, we have a trace propagated between the Datadog-instrumented API and the
> OpenTelemetry-instrumented API, however, all child spans generated by Datadog

Reviewer: ```suggestion
Looking at the Jaeger UI, you can see that the problem was partially solved.
Now you have a trace propagated between the Datadog-instrumented API and the
OpenTelemetry-instrumented API. However, all child spans generated by Datadog
```

are segregated in a different trace, defined only as the Lower TraceID. This
happens because `ddtrace` only sends a `_dd.p.tid` attribute for the first span

> generated internally, which makes our transform statements skip the child spans.

Reviewer: ```suggestion
generated internally, which makes your transform statements skip the child spans.
```

![Partially disjointed OTelSDK-instrumented trace on Jaeger](./step2-otel-api.png)

![Partially disjointed Datadog-instrumented trace on Jaeger](./step2-datadog-api.png)

> ### Patching Datadog Trace to Send the Upper TraceID to Every Child Span

Reviewer: ```suggestion
### Patch Datadog trace to send the Upper TraceID to every child span
```

Since the OpenTelemetry Collector was
[designed to process spans considering distributed systems](https://opentelemetry.io/docs/collector/scaling/),

> we don't have a way to maintain an internal state to replicate the `_dd.p.tid`

Reviewer: ```suggestion
there isn't have a way to maintain an internal state to replicate the `_dd.p.tid`
```

attribute for the child spans received by the `datadog` receiver.

> We will solve this problem directly on the Datadog-instrumented API, applying a
> minor patch to `ddtrace` to replicate `_dd.p.tid` to all child spans. In the
> `ddtrace` Ruby version, we were able to do that by changing the trace
> serialization to send this data as shown
> [here](https://github.com/kubeshop/tracetest/blob/main/examples/datadog-propagation/quick_start_api_datadog/config/initializers/datadog.rb#L20):

Reviewer: ```suggestion
You can solve this problem directly on the Datadog-instrumented API by applying a
minor patch to `ddtrace` to replicate `_dd.p.tid` to all child spans. In the
[`ddtrace` Ruby version](https://github.com/kubeshop/tracetest/blob/main/examples/datadog-propagation/quick_start_api_datadog/config/initializers/datadog.rb#L20), the trace
serialization is modified to send this data as shown:
```

```ruby
module Datadog
  module Tracing
    module Transport
      class SerializableTrace
        def to_msgpack(packer = nil)
          if ENV.has_key?('INJECT_UPPER_TRACE_ID')
            return trace.spans.map { |s| SerializableSpan.new(s) }.to_msgpack(packer)
          end

          upper_trace_id = trace.spans.find { |span| span.meta.has_key?('_dd.p.tid') }.meta['_dd.p.tid']
          trace.spans.each do |span|
            span.meta["propagation.upper_trace_id"] = upper_trace_id
          end

          trace.spans.map { |s| SerializableSpan.new(s) }.to_msgpack(packer)
        end
      end
    end
  end
end
```

> Since it is a customization on the traces, we opted to grab the `_dd.p.tid`
> attribute and inject it in each span as the `propagation.upper_trace_id`
> attribute. We
> [changed](https://github.com/kubeshop/tracetest/blob/main/examples/datadog-propagation/collector/collector.config.step3.yaml#L18)

Reviewer: ```suggestion
Since it is a customization on the traces, you can opt to grab the `_dd.p.tid`
attribute and inject it in each span as the `propagation.upper_trace_id`
attribute. Then you can
[change](https://github.com/kubeshop/tracetest/blob/main/examples/datadog-propagation/collector/collector.config.step3.yaml#L18)
```

the `transform` processor in the OpenTelemetry Collector to consider this:

```yaml
transform:
  trace_statements:
    - context: span
      statements:
        - set(cache["upper_trace_id"], attributes["propagation.upper_trace_id"])
          where attributes["propagation.upper_trace_id"] != nil
        - set(cache["lower_trace_id"], Substring(trace_id.string, 16, 16)) where
          cache["upper_trace_id"] != nil
        - set(cache["combined_trace_id"], Concat([cache["upper_trace_id"],
          cache["lower_trace_id"]],"")) where cache["upper_trace_id"] != nil
        - set(trace_id.string, cache["combined_trace_id"]) where
          cache["combined_trace_id"] != nil
```

With these changes done, let’s run a new setup with our example:

```bash
> docker compose -f ./docker-compose.step3.yaml up -d

> > curl http://localhost:8080/remotehello
# it returns:
# {"hello":"world!","from":"quick-start-api-otel","to":"quick-start-api-datadog"}
```

> Opening the Jaeger UI again we see that our problem was solved! Now we have a

Reviewer: ```suggestion
Opening the Jaeger UI again, you see that the problem is solved! Now you have a
```

single trace between both APIs and can evaluate a process as a whole in a
developer machine.

![Unified OTelSDK-instrumented and Datadog-instrumented trace on Jaeger](./step3.png)

> ## Final Remarks

Reviewer: ```suggestion
## Final remarks
```

We have discussed integrating Datadog-instrumented apps in an OpenTelemetry
stack, seeing the problem of different TraceID representations for Datadog and

> OpenTelemetry, which can lead to disjointed traces. To solve that we are using
> an OpenTelemetry Collector with a transform processor to reconstruct the TraceID

Reviewer: ```suggestion
OpenTelemetry, which can lead to disjointed traces. To solve that we used
an OpenTelemetry Collector with a `transform` processor to reconstruct the TraceID
```

for Datadog spans, facilitating a unified tracing environment.

As a team focused on building an open source tool in the observability space,

> the opportunity to help the overall OpenTelemetry community was important to us.
> That’s why we were researching and finding new ways of collecting traces from

Reviewer: ```suggestion
the opportunity to help the overall OpenTelemetry community is important to us.
That’s why we are researching and finding new ways of collecting traces from
```
